src/foam/nanos/auth/ServiceProvider.js

```diff
@@ -34,5 +49,102 @@ foam.CLASS({
+              EQ(UserCapabilityJunction.TARGET_ID, capability.getId())
+            ));
+            if ( ucj != null ) 
+              ucj = (UserCapabilityJunction) ucj.fclone();
+            else 
+              ucj = new UserCapabilityJunction.Builder(x).setSourceId(user.getId()).setTargetId(capability.getId()).build();
+            ucj.setStatus(CapabilityJunctionStatus.GRANTED);
+            ucj = (UserCapabilityJunction) userCapabilityJunctionDAO.put_(x, ucj);
+            if ( ucj == null || ucj.getStatus() !=foam.nanos.crunch.CapabilityJunctionStatus.GRANTED )
+              throw new RuntimeException("Error setting up UserCapabilityJunction for user: " + user.getId() + " and spid: " + getId());
+          }
+        } catch (Exception e) {
+          logger.warning(e);
+        }
+      `
+    },
+    {
+      name: 'removeSpid',
+      args: [
+        { name: 'x', javaType: 'foam.core.X' },
+        { name: 'user', javaType: 'foam.nanos.auth.User' }
+      ],
+      documentation: `
+        Method used for removing a user's old ServiceProvider capability and any capability that is a part
+        of that capability through prerequisite junctions.
+        Called before a user is assigned a new ServiceProvider capability
+      `,
+      javaCode: `
+        CrunchService crunchService = (CrunchService) x.get("crunchService");
+        DAO userCapabilityJunctionDAO = (DAO) x.get("bareUserCapabilityJunctionDAO");
+
+        // find list of old spids to remove from user
+
+        List<UserCapabilityJunction> spidsToRemove = new ArrayList<UserCapabilityJunction>();    
+        AbstractPredicate serviceProviderTargetPredicate = new AbstractPredicate(x) {
+          @Override
+          public boolean f(Object obj) {
+            UserCapabilityJunction ucJunction = (UserCapabilityJunction) obj;
+            Capability c = (Capability) ucJunction.findTargetId(x);
+            return c instanceof ServiceProvider;
+          }
+        };
+        spidsToRemove = (ArrayList<UserCapabilityJunction>) ((ArraySink) userCapabilityJunctionDAO
+          .where(AND(
+            EQ(UserCapabilityJunction.SOURCE_ID, user.getId()),
+            NEQ(UserCapabilityJunction.TARGET_ID, getId()),
+            serviceProviderTargetPredicate
+          )).select(new ArraySink())).getArray();
+
+        // for each old spid, get its capabilityPath, and remove all ucjs on the capabilityPath
+
+        for ( UserCapabilityJunction sp : spidsToRemove ) {
+          List<Capability> capabilitiesToRemove = (List<Capability>) crunchService.getCapabilityPath(x, sp.getTargetId(), false);
+          AbstractPredicate removePredicate = new AbstractPredicate(x) {
+            @Override
+            public boolean f(Object obj) {
+              UserCapabilityJunction ucJunction = (UserCapabilityJunction) obj;
+              Capability c = (Capability) ucJunction.findTargetId(x);
+              return capabilitiesToRemove.contains(c);
+            }
+          };
```

Comment:
Can we use IN(capabilitiesToRemove.ids) instead?

Reply:
would that work with list of Capability in a predicate for usercapabilityjunction?

Reply:
could be IN(Ucj.TARGET_ID, capabilitiesToRemove.ids)

Reply:
okok i see now, will update, thanks